src/ccder.c

```diff
@@ -141,62 +302,161 @@ uint8_t *ccder_encode_eckey(size_t priv_size, const uint8_t *priv_key,
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t body_len = 0;
+	const uint8_t* body_start = ccder_decode_tl(expected_tag, &body_len, der, der_end);
+	*body_end = body_start + body_len;
+	return body_start;
 }
 
 const uint8_t *
 ccder_decode_sequence_tl(const uint8_t **body_end,
                          const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	// NOTE(@facekapow):
+	// we don't just call `ccder_decode_constructed_tl` because in the future,
+	// that method will check that the decoded tag is a CONSTRUCTED tag
+
+	size_t body_len = 0;
+	const uint8_t* body_start = ccder_decode_tl(CCDER_SEQUENCE, &body_len, der, der_end);
+	*body_end = body_start + body_len;
+	return body_start;
 }
 
 const uint8_t *ccder_decode_uint_n(cc_size *n, 
                                  const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t num_len = 0;
+	const uint8_t* num_start = ccder_decode_tl(CCDER_INTEGER, &num_len, der, der_end);
+	while (num_start[0] == 0) {
+		--num_len;
+		++num_start;
+	}
+	*n = ccn_nof_size(num_len);
+	return num_start + num_len;
 }
 
 const uint8_t *ccder_decode_uint(cc_size n, cc_unit *r,
                                  const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t num_len = 0;
+	const uint8_t* num_start = ccder_decode_tl(CCDER_INTEGER, &num_len, der, der_end);
+	while (num_start[0] == 0) {
+		--num_len;
+		++num_start;
+	}
+	ccn_read_uint(n, r, num_len, num_start);
+	return num_start + num_len;
 }
 
 const uint8_t *ccder_decode_uint64(uint64_t* r,
                                    const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t num_len = 0;
+	const uint8_t* num_start = ccder_decode_tl(CCDER_INTEGER, &num_len, der, der_end);
+	while (num_start[0] == 0) {
+		--num_len;
+		++num_start;
+	}
+
+	*r = 0;
+
+	for (size_t i = 0; i < num_len; ++i) {
+		*r |= (uint64_t)num_start[num_len - 1 - i] << (i * 8);
+	}
+
+	return num_start + num_len;
 }
 
 const uint8_t *ccder_decode_seqii(cc_size n, cc_unit *r, cc_unit *s,
                                   const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t seq_len = 0;
+	return ccder_decode_uint(n, s, ccder_decode_uint(n, r, ccder_decode_tl(CCDER_SEQUENCE, &seq_len, der, der_end), der_end), der_end);
 }
+
 const uint8_t *ccder_decode_oid(ccoid_t *oidp,
                                 const uint8_t *der, const uint8_t *der_end) {
-	printf("DARLING CRYPTO STUB: %s\n", __PRETTY_FUNCTION__);
+	size_t oid_len = 0;
+	const uint8_t* oid_start = ccder_decode_tl(CCDER_OBJECT_IDENTIFIER, &oid_len, der, der_end);
+	size_t total_len = (size_t)(oid_start - der) + oid_len;
+
+	uint8_t* copy = malloc(sizeof(uint8_t) * total_len + 1);
```

Comment:
Is the use of `malloc` here and in `ccder_decode_bitstring` still ok? We can't use `alloca` because we need to return the values.

This is assuming, of course, that the original Apple API was supposed to create copies and not just return pointers to the values in the DER structure. I couldn't really find any uses of the DER encoding/decoding API in open source software (Google only returned results for the copyrighted Apple CoreCrypto source, and we can't look at that for obvious reasons) and there's no header documentation either.

Reply:
(I'm not a lawyer, but) you can look at Apple's CoreCrypto, you just cannot, like, copy-paste from there (or write virtually the same code yourself). But you can look at it to figure out whether it mallocs returned memory or what.

Reply:
Hmmm. Is it possible to use the memory between `der` and `der_end` for the return?

Usually what they do in this case is use some of the memory that was passed in.

Reply:
(or at least some of the memory, then return a pointer to it's start)

Reply:
We can do this in a clean-room way, if you want. You can ask questions and we give answers.

No memory is allocated in this function. The pointer it gets from ccder_decode_tl() is expected to be persistent.

Reply:
I'm okay with looking at Apple's code in order to see how the API is meant to be used and such, I just don't want to subconsciously copy the code without realizing it. So, yes, I'll try to do this clean-room as much as I can.

Besides, the only real reason I'd have for checking Apple's code is stuff like this (i.e. internals that might affect external behavior, e.g. where I'm not sure if they allocate memory or not). For the most part, the APIs are pretty simple (once you consider RFCs/standards).

@LubosD Thanks for the info, I'll update it.